Python Record_Component: make_constant

ax3l opened the pull request:
Refactor numpy detection of Attributable to general method. Reuse in Record_Component bindings for make_constant.

Also fixes the lifetime of `Container` members (`iterations`, `meshes`, `particles`) in `Series` and `Iteration`. Returned (referenced) containers need to be freed before their owner classes.

- [x] depends on #358

ax3l committed:
Python Record_Component: make_constant

Refactor numpy detection of Attributable to general
method. Reuse in Record_Component bindings for make_constant.

anokfireball reviewed (approved):
Unit tests would be nice down the line.

ax3l committed:
Python Test: make_constant

ax3l commented on src/binding/python/RecordComponent.cpp:
```diff
@@ -51,20 +51,86 @@ void init_RecordComponent(py::module &m) {
+        }, py::arg("value")
+        )
+        // allowed python intrinsics, after (!) buffer matching
+        .def("make_constant", &RecordComponent::makeConstant<char>,
```
must be `std::string`: Python has no native `char`s and otherwise our strings are all casted to size-one char

ax3l replied:
but a `std::string` makes no real sense as a value for a record. so let's keep `char`s here.

ax3l commented:
@C0nsultant unit test added as well :)

ax3l commented on test/python/unittest/API/APITest.py:
```diff
@@ -308,6 +308,121 @@ def testAttributes(self):
+                                                      extent))
+            ms["longdouble"][SCALAR].make_constant(np.longdouble(1.23456789))
+
+        # flush and close file, TODO affected by #357
```
seems to be affected by #357 in both HDF5 and ADIOS1

ax3l commented on test/python/unittest/API/APITest.py:
```diff
@@ -308,6 +308,122 @@ def testAttributes(self):
+            openPMD.Access_Type.create
+        )
+
+        ms = series.iterations[0].meshes
```
That object needs love via [return policies](https://pybind11.readthedocs.io/en/stable/advanced/functions.html). Using that shortcut variable corrupts things.

ax3l replied:
Fixed: had to ensure returned `Container`s are freed before the object that exposed them in the garbage collector

ax3l committed:
Python: Container Member Lifetimes

Make sure the garbage collector frees public container members before
their owner classes (Series, Iteration).
